src/style.js

```diff
@@ -244,9 +244,13 @@ function groupAesthetics({
   return [AL, T, F, FO, S, SO, SW, O, H].filter((c) => c !== undefined);
 }
 
+export function isHighCardinality(value) {
+  return new InternSet(value).size > value?.length >> 1;
+}
+
 export function groupZ(I, Z, z) {
   const G = group(I, (i) => Z[i]);
-  if (z === undefined && G.size > I.length >> 1) {
+  if (z === undefined && isHighCardinality(I.map((i) => Z[i]))) {
```

Comment:
We should leave this line as-is; reusing the isHighCardinality function here is extra work (because we have to create an InternSet when we’ve already done it with `G`).